csrc/logical_domain_map.cpp

```diff
@@ -284,19 +284,25 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseLogicalDomainMap::map(
     bool consumer_has_e = consumer_tv_->sameAs(op->grad_query()) ||
         consumer_tv_->sameAs(op->grad_key());
 
+    size_t num_device_dim =
+        producer_logical.size() > 0 && producer_logical.at(0)->isDeviceDim()
+        ? 1
+        : 0;
     for (auto idx : c10::irange(producer_logical.size())) {
-      if (idx < 2) {
+      if (idx < (2 + num_device_dim)) {
         // Map N, H from all producers to consumers
         updatePairwiseLogicalDomainMap(
             producer_logical.at(idx), consumer_root.at(idx));
-      } else if (idx == 2 && (producer_has_s == consumer_has_s)) {
+      } else if (
+          idx == (2 + num_device_dim) && (producer_has_s == consumer_has_s)) {
```

Comment:
Parens are not used consistently. I'd do
```suggestion
          idx == 2 + num_device_dim && producer_has_s == consumer_has_s) {
```
or if you much prefer parens

```
          (idx == 2 + num_device_dim) && (producer_has_s == consumer_has_s)) {
```